tests/5.0/atomic/test_atomic_num_hint.F90

```diff
@@ -0,0 +1,80 @@
+!//===------ test_atomic_num_hint.F90 -------------------------------------===//
+!
+! OpenMP API Version 5.0 Nov 2018
+!
+!  This test checks if atomic hints passed by enum value
+!  are accepted by the compiler. If the sync hint is not
+!  yet defined in the specification, it defaults to 
+!  omp_sync_hint_none (0x0). 
+!
+!//===---------------------------------------------------------------------===//
+
+#include "ompvv.F90"
+
+#define N 1024
+
+PROGRAM test_atomic_num_hint
+  USE iso_fortran_env
+  USE ompvv_lib
+  USE omp_lib
+  implicit none
+
+  OMPVV_TEST_VERBOSE(test_atomic_with_used_enum_value() .ne. 0)
+  OMPVV_TEST_VERBOSE(test_atomic_with_unused_enum_value() .ne. 0)
+
+  OMPVV_REPORT_AND_RETURN()
+
+CONTAINS
+  INTEGER FUNCTION test_atomic_with_used_enum_value()
+    INTEGER:: x, num_threads, errors
+
+    errors = 0
+    x = 0
+    num_threads = -1
+
+    OMPVV_INFOMSG("test_atomic_with_used_enum_value")
+
+    !$omp parallel num_threads(2) default(shared)
+       IF (omp_get_thread_num() .EQ. 0) THEN
+          num_threads = omp_get_num_threads()
+       END IF
+       !$omp atomic hint(4) ! corrosponds to omp_sync_hint_nonspeculative
+          x = x + 1
+       !$omp end atomic 
+    !$omp end parallel
+
+    OMPVV_ERROR_IF(num_threads .LT. 0, "Test ran with invalid number of threads (less than zero)")
+    OMPVV_WARNING_IF(num_threads .EQ. 1, "Test ran with one thread, so the results are not conclusive")
+
+    OMPVV_TEST_AND_SET_VERBOSE(errors, x .NE. num_threads)
+
+    test_atomic_with_used_enum_value = errors
+  END FUNCTION test_atomic_with_used_enum_value
+
+  INTEGER FUNCTION test_atomic_with_unused_enum_value()
+    INTEGER:: x, num_threads, errors
+
+    errors = 0
+    x = 0
+    num_threads = -1
+
+    OMPVV_INFOMSG("test_atomic_with_unused_enum_value")
+
+    !$omp parallel num_threads(OMPVV_NUM_THREADS_HOST) default(shared)
+       IF (omp_get_thread_num() .EQ. 0) THEN
+          num_threads = omp_get_num_threads()
+       END IF
+       !$omp atomic hint(4132) ! As of OMP Spec 5.0 only values till 0x8 have been taken
```

Comment:
(BTW, the highest possible value is not 8 but 8 + 4 + 2 + 1 =15.)

I wonder what's the purpose to use a non-taken number here; what's the expected result?

While I see the wording _"The valid constants must include the following, which can be extended with implementation-defined values:"_ which permits additional values, it is unclear what an implementation should/might do with a wrong value: ignore? compile-time error? runtime-error?

I assume that many implementations fully ignore the _hint_ (after parsing) – at least GCC currently does so.

Reply:
As I understand, spec does not clarify what should happen when non-taken number is passed as a hint, but spec says that "The presence of a hint clause does not a ffect the semantics of the atomic construct, and all hints may be ignored.", and thus the purpose of this test may be to check whether the implementation CAN ignore the unexpected hint argument.

Reply:
By the way, "As of OMP Spec 5.0 only values till 0x8 have been taken" means that OpenMP V5.0 has defined up to 0x8 bit (and the highest possible value is not 15 because there are restrictions on which hint values can be combined).